compiler/noirc_evaluator/src/ssa/opt/inlining.rs

```diff
@@ -81,47 +89,47 @@ impl Function {
     pub(super) fn inlined(
         &self,
         ssa: &Ssa,
+        should_inline_call: &impl Fn(&PerFunctionContext, &Ssa, FunctionId) -> bool,
+    ) -> Function {
+        InlineContext::new(ssa, self.id()).inline_all(ssa, &should_inline_call)
+    }
+
+    /// Generic function that determines whether a function should inline a call.
+    pub(super) fn should_inline_call(
+        &self,
+        ssa: &Ssa,
+        called_func_id: FunctionId,
         inline_no_predicates_functions: bool,
         inline_infos: &InlineInfos,
-    ) -> Function {
-        let caller_runtime = self.runtime();
-
-        let should_inline_call =
-            |_context: &PerFunctionContext, ssa: &Ssa, called_func_id: FunctionId| -> bool {
-                // Do not inline self-recursive functions on the top level.
-                // Inlining a self-recursive function works when there is something to inline into
-                // by importing all the recursive blocks, but for the entry function there is no wrapper.
-                if called_func_id == self.id() {
-                    return false;
-                }
-                let callee = &ssa.functions[&called_func_id];
+    ) -> bool {
+        // Do not inline self-recursive functions on the top level.
+        // Inlining a self-recursive function works when there is something to inline into
+        // by importing all the recursive blocks, but for the entry function there is no wrapper.
+        if called_func_id == self.id() {
+            return false;
+        }
```

Comment:
Note that these checks should be enforced even if we're passing a different inlining condition in as the closure as theyre important safety checks. Can you pull them out of this function and ensure they're always enforced?